Background: add background attachment to top level styles

ramonjd opened the pull request:
## What?

Adds a toggle to allow background images to scroll with the page, or remain fixed in position.

Part of:

- https://github.com/WordPress/gutenberg/issues/54336


Core backport PR:

- https://github.com/WordPress/wordpress-develop/pull/6991

## Why?

To increase the possibilities of background images.

## How?

Adding support for `background-attachment`.

## TODO

- [x] Add backport PR
- [x] Get advice on placement/design

## Testing Instructions

1. Insert a Group or Verse block.
2. Add a background image to the block
3. Check that image scrolls with the page by default
4. Toggle the "Scroll with page" option off.
5. Test that the background image is now fixed.
6. Save and check the frontend
7. Repeat with a site background image in the Site Editor (under Styles > Layout) 

### Testing Instructions for Keyboard


1. Ensure you can tab through the sidebar controls to the background attachment toggle
2. To activate/deactivate hit the space bar


## Screenshots or screencast <!-- if applicable -->





https://github.com/WordPress/gutenberg/assets/6458278/8067a8ce-f2ad-4101-a470-33bae4dd6015

ramonjd committed:
Merge with control panel.

ramonjd commented on packages/block-editor/src/components/global-styles/background-panel.js:
```diff
@@ -552,6 +564,14 @@ function BackgroundSizeControls( {
 				value={ backgroundPositionToCoords( positionValue ) }
 				onChange={ updateBackgroundPosition }
 			/>
+			<ToggleControl
```
I've put no thought put into this location or the actual control. Could you tell? 🤣 

It does affect the height of the popover, so maybe there's some optimization required.

Elsewhere I've tried using a select dropdown for the size options, and having it site to the side, e.g.,

<img width="405" alt="Screenshot 2024-07-05 at 11 47 45 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/32cb0d09-f9dd-4148-9fc0-19351bb014cf">

It's a bit of a control salad 😄 

The problem is that translations can and do push out the copy.

I wonder if an accordion or something similar might help to use the space more evenly?

The help text for this toggle could be removed as well.

ramonjd committed:
LINTY

jameskoster commented:
Grouping the toggle with the position UI makes sense to me. But we should probably align more closely with the equivalent control in the Cover block, at least initially:

<img width="296" alt="Screenshot 2024-07-05 at 12 02 32" src="https://github.com/WordPress/gutenberg/assets/846565/f881464f-7e59-44f5-8056-9f7d42a84e86">

Specifically when the toggle is _off_, `background-attachment` should be `scroll`. Perhaps the label just needs to be 'Fixed'?

---

Obviously this is separate, but one thing we could try to streamline the panel a bit is hide the position inputs behind a `settings` toggle button:

<img width="302" alt="Screenshot 2024-07-05 at 12 05 14" src="https://github.com/WordPress/gutenberg/assets/846565/d77bc392-d282-4f13-a209-f0f6f7037a22">

ramonjd committed:
Change the background attachment toggle label to "Fixed background" and flip the checked condition

ramonjd commented:
> Grouping the toggle with the position UI makes sense to me. But we should probably align more closely with the equivalent control in the Cover block, at least initially:

G	ood one. Will do.

> but one thing we could try to streamline the panel a bit is hide the position inputs behind a settings toggle button:

Yeah, that sounds like a worth experiment. Thanks for the idea!

ramonjd committed:
Add background attachment block test in theme json

ramonjd commented:
> > but one thing we could try to streamline the panel a bit is hide the position inputs behind a settings toggle button:


Just to confirm, something similar to how the font size controls behave:

https://github.com/WordPress/gutenberg/assets/6458278/dbf67ad1-1636-4b00-a37b-65dee8963dcb

ramonjd committed:
LINTY McLINTFACE

ramonjd committed:
Add backport changelog

ramonjd committed:
Add style engine test

Add style engine test

jameskoster commented:
Similar, but clicking the settings icon would toggle the visibility of the position inputs (additional to the focal point picker).

aaronrobertshaw commented on packages/block-editor/src/components/global-styles/background-panel.js:
```diff
@@ -451,6 +451,9 @@ function BackgroundSizeControls( {
 		inheritedValue?.background?.backgroundPosition;
+	const attachmentValue = style?.background?.backgroundAttachment || {
+		...inheritedValue?.background?.backgroundAttachment,
+	};
```
```suggestion
	const attachmentValue =
		style?.background?.backgroundAttachment ||
		inheritedValue?.background?.backgroundAttachment;
```

Am I correct in thinking the `backgroundAttachment` is a simple string and the inherited value should be used directly?

If I apply the suggestion locally, it fixes the control not displaying the inherited value from theme.json.

Should we also be updating the theme.json schema to indicate the expected type for `backgroundAttachment`, or is that premature?

ramonjd replied:
You're right on all counts. I've pushed a fix. 👍🏻

aaronrobertshaw reviewed (commented):
Nice work @ramonjd 👍 

This is testing pretty well for me so far. 

✅ Default attachment is scroll
✅ Toggling background attachment works as expected in the editor
✅ Resetting control clears the attachment selection

I did run into a couple of issues though when testing the application of `backgroundAttachment` via theme.json.

1. If I set the `backgroundAttachment` value in theme.json it isn't reflected in the background panel control. (should be an easy fix 🤞 )
2. With a theme.json `backgroundAttachment` value of `fixed`, if I toggle off the control in the editor for a specific block instance, it works there (scroll) but the frontend remains fixed.

This is what I am seeing (please excuse dodgy test content):


https://github.com/WordPress/gutenberg/assets/60436221/59175dca-d2f9-4960-b32e-95341bdb68d8


The absence of a background attachment value on the block instance means no style is applied for the block instance and it falls back to the global style for the block type.

<img width="470" alt="Screenshot 2024-07-10 at 6 07 17 PM" src="https://github.com/WordPress/gutenberg/assets/60436221/0651b7a1-124d-4e9a-ac9e-7e297d043819">


The last thing I noticed is unrelated to this PR but might be good to address in follow-ups is the inconsistent positioning of the background image panel in the inspector sidebar.

| Group | Verse |
|---|---|
| <img width="267" alt="Screenshot 2024-07-10 at 5 59 31 PM" src="https://github.com/WordPress/gutenberg/assets/60436221/c83ade96-906a-4098-b02c-4c5a1b6f3dc8"> | <img width="282" alt="Screenshot 2024-07-10 at 5 59 37 PM" src="https://github.com/WordPress/gutenberg/assets/60436221/98e99f7c-0a59-4352-83dc-da1f3909d47b"> |

ramonjd committed:
update theme.json schema
fix wrong value assignment

ramonjd committed:
Pass resolved links to the component

ramonjd commented:
> If I set the backgroundAttachment value in theme.json it isn't reflected in the background panel control. (should be an easy fix 🤞 )

Ah, do you mean with relative paths from a block style? I could reproduce that. I'd forgotten to pass the resolved theme paths to the component in https://github.com/WordPress/gutenberg/pull/60100

I've just pushed a fix.

> With a theme.json backgroundAttachment value of fixed, if I toggle off the control in the editor for a specific block instance, it works there (scroll) but the frontend remains fixed.

I think that was related to the bug you found [above](https://github.com/WordPress/gutenberg/pull/61382#discussion_r1671805091) if I'm not mistaken. Good spotting! That was a copy past face palm disaster. 

 I've updated with your suggested change. Thanks!

I also updated the theme.json schema with `backgroundAttachment`, which I'd forgotten to do.

Thanks for the thorough testing @aaronrobertshaw 

Feel free to push to this branch and merge if you think it's okay.

ramonjd commented:
> The last thing I noticed is unrelated to this PR but might be good to address in follow-ups is the inconsistent positioning of the background image panel in the inspector sidebar.


Yes, good point. It should be consistent between block styles and block global styles.

aaronrobertshaw reviewed (approved):
Thanks for iterating on this one @ramonjd 👍 

I can confirm the two issues I noted in my last review have been fixed and this is testing nicely for me.

While testing I noticed that display of the background-attachment control cannot be set via theme.json settings. It is also conditional upon background-size being enabled, which might be valid given its display within the background-size popover panel. It's display within that popover should be controllable by theme authors though.

It might be best to address the toggling of the control in a follow-up as I don' t think it needs to block this PR.

> Yes, good point. It should be consistent between block styles and block global styles.

The screenshots in my review weren't even between the block inspector and global styles panels. Only the block inspector for Verse vs Group blocks. I'll see what I can do to make them consistent across the board.


---

@jameskoster would you be okay with the [proposed UI update](https://github.com/WordPress/gutenberg/pull/61382#issuecomment-2210680981) for the background-attachment control to be exposed via a settings toggle, to be explored in a follow-up? 

Also, do you think the background-attachment control should be enabled within `appearanceTools`?

aaronrobertshaw commented:
I've put up a PR to locate the background image panel consistently: https://github.com/WordPress/gutenberg/pull/63551

jameskoster commented:
> would you be okay with the https://github.com/WordPress/gutenberg/pull/61382#issuecomment-2210680981 for the background-attachment control to be exposed via a settings toggle, to be explored in a follow-up?

Yes, but to be clear, only the X/Y position need to be hidden since they're effectively duplicative of the focal point picker. The control for fixing the bg should probably be permanently visible.

aaronrobertshaw commented:
I've started looking into adding a toggle to the `FocalPointPicker` to hide/show the additional position controls. I think this is going to be a more significant refactor of the component than anticipated. 

The `FocalPointPicker` currently has a `BaseControl` wrapper that won't easily accept a toggle button into the label/header. 

@ciampo or @mirka can you share any thoughts on the best approach to achieving [this](https://github.com/WordPress/gutenberg/pull/61382#issuecomment-2228252898)? Does it warrant a v2 of the component?

ciampo commented:
@aaronrobertshaw would something like [this](https://github.com/WordPress/gutenberg/pull/63742) work, or am I misunderstanding the request?

mirka commented:
I wonder if it's time to add a `labelSuffix` prop to `BaseControl`. It's clearly a pattern at this point (e.g. FontSizePicker, SpacingSizesControl, probably more), but they're being implemented ad hoc.

aaronrobertshaw commented:
> would something like https://github.com/WordPress/gutenberg/pull/63742 work, or am I misunderstanding the request?

Thanks, that part of this is straightforward. The problem is we need a button within the `FocalPointPicker` component, level with, or a suffix to, the label to toggle the display of those controls. 

> I wonder if it's time to add a labelSuffix prop to BaseControl

Given how widespread the use of `BaseControl` is I wasn't too sure if there would be much appetite for messing with it. FWIW I think it is a pattern that is going to continue as our UI grows more complex and we need finer-grained control.

ramonjd commented:
>  The problem is we need a button within the FocalPointPicker component, level with, or a suffix to, the label to toggle the display of those controls.

I haven't looked too deeply at the FocalPointPicker component, but would it be acceptable to add a prop that would render the "toggling version", e.g., allowing hide/unhide of [FocalPointPickerControls](https://github.com/WordPress/gutenberg/blob/cd8d59fd351ba1a3e827218abede4c9086aae1c1/packages/components/src/focal-point-picker/controls.tsx#L24-L24)?

I'm not great at naming, except maybe for small household pets...

```jsx

<FocalPointPicker pointControls="toggle" /> { /* maybe "hidden" as well? */ }

```

ciampo commented:
> > The problem is we need a button within the FocalPointPicker component, level with, or a suffix to, the label to toggle the display of those controls.
> 
> I haven't looked too deeply at the FocalPointPicker component, but would it be acceptable to add a prop that would render the "toggling version", e.g., allowing hide/unhide of [FocalPointPickerControls](https://github.com/WordPress/gutenberg/blob/cd8d59fd351ba1a3e827218abede4c9086aae1c1/packages/components/src/focal-point-picker/controls.tsx#L24-L24)?
> 
> I'm not great at naming, except maybe for small household pets...
> 
> ```js
> <FocalPointPicker pointControls="toggle" /> { /* maybe "hidden" as well? */ }
> ```

@ramonjd from what I understand, the tricky part is not hiding/showing the controls, but to add an interactive control, next to the label, that the user can use to toggle such input controls on and off.

Currently, given how the component works, that is only possible by adding such toggle externally to the component — folks are wondering if we could do better.

mirka commented:
I'm ok with adding a `labelSuffix` prop to BaseControl (unless we discover unexpected blockers), as well as a new prop on FocalPointPicker.
